### PR TITLE
Clip Binomial results for different endpoints in curand_uniform

### LINKED ISSUE #42153: [bug] Binomial distribution has small chance of returning -1
## 🐛 Bug

As it says in the title, I found that the binomial sampler can occasionally return `-1` (about 1 in a 100M trials). This is probably mostly a small issue, but it's a big issue if you're computing log probabilities from the output.

I created PR #38456 for this, which was then merged. Unfortunately the bug still appears which means that my fix was not relevant (I was not able to build pytorch at the time for testing, that's my bad!). I think that PR can remain merged as its harmless, but a fix for the bug is still needed.

Looking at the code again, my fix was obviously wrong because the repro case below doesn't even go through the BTRS code path 🤦

<details>
<summary>edit: this part was wrong, see comments below</summary>
I believe the bug is in [`binomial_inversion`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Distributions.h#L159-L177), specifically line 169. If the result of the standard uniform is very close to 1 (like, `> 0.9999999`) then `log(U)` evaluates to 0 and `geom_sum` doesn't increase for a round. The rate of this matches pretty closely with the rate I observe the bug, which is comforting because last time I had no friggin' idea why it was happening.

edit to add: in my previous PR I said this was only on CUDA, but I think I just wasn't patient enough to sample enough on the CPU. Now I see it there as well.
</details>

## To Reproduce

This reliably returns `False`:

```python
(
    torch.distributions.Binomial(
        total_count=torch.tensor(1.0).cuda(), probs=torch.tensor(0.9).cuda()
    ).sample(torch.Size((1000000000,))) >= 0
).all()
```

This also reliably returns `False`:

```python
(
    torch.distributions.Binomial(
        total_count=torch.tensor(1.0).cuda(), probs=torch.tensor(0.1).cuda()
    ).sample(torch.Size((1000000000,))) < 2
).all()
```

(see [Bender's nightmare](https://www.youtube.com/watch?v=MOn_ySghN2Y))

## Expected behavior

Should return `True`!

## Environment

 - PyTorch Version (e.g., 1.0): Running pytorch-nightly 1.7.0.dev20200713
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: Titan Xp 
 - Any other relevant information:

cc @ezyang @gchanan @zou3519 @vincentqb @fritzo @neerajprad @alicanb @vishwakftw @pbelevich

### PR DESCRIPTION
**Pull request by mattip**
Fixes #42153

As [documented](https://docs.nvidia.com/cuda/curand/device-api-overview.html) (search for `curand_uniform` on the page), `curand_uniform` returns "from 0.0 to 1.0, where 1.0 is included and 0.0 is excluded." These endpoints are different than the CPU equivalent, and makes the calculation in the PR fail when the value is 1.0.

The test from the issue is added, it failed for me consistently before the PR even though I cut the number of samples by 10.

### CODE REVIEW

**Comment by ailzhang** [comment]
cc @ngimel for a review since you are in the original issue thread, please feel free to remove the assignment if doesn't fit.

**Comment by mattip** [comment]
Lint error is due to CMake not finding Protobuf:
```
CMake Error at /usr/local/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Could NOT find Protobuf: Found unsuitable version
  "Protobuf_VERSION_NOTFOUND", but required is at least "3" (found
  Protobuf_LIBRARY-NOTFOUND;-pthread)
```

caffe2_onnx_main_py3_6_clang7_ubuntu16_04_build failure is caused by a CMake version discrepancy"
```
CMake Error at third_party/tensorpipe/CMakeLists.txt:12 (cmake_minimum_required):
Aug 06 22:05:47   CMake 3.17 or higher is required.  You are running version 3.5.1
```

I don't think I caused those ...

File: aten/src/ATen/native/Distributions.h [hunk 1]
```diff
@@ -166,7 +166,13 @@ C10_DEVICE scalar_t binomial_inversion(scalar_t count, scalar_t prob, BaseSample
 
   while (1) {
     U = standard_uniform.sample();
-    accscalar_t geom = compat_ceil(compat_log(U) / logprob);
+    // Cuda's curand_uniform can return 1.0 but not 0.0
+    // Reject 1.0 since it will produce geom == 0.0
+    accscalar_t logU = compat_log(U);
+    if (logU == 0.0) {
```
[the remark this answers is not in the document]
Comment (mattip) [reply aten/src/ATen/native/Distributions.h:172]: Done. I was hesitant to do such a sweeping change at first in fear it might change values for people who use a seed and expect consistent results. Is there a stream-constant guarantee in PyTorch?

**Comment by mattip** [comment]
> I don't think I caused those ...

Yes, I did. I picked up some third-party submodule changes by mistake. Fixing.

File: aten/src/ATen/native/cuda/Distributions.cu [hunk 2]
```diff
@@ -69,7 +69,13 @@ struct curand_uniform_wrapper {
   curandStatePhilox4_32_10_t &state;
   __device__ curand_uniform_wrapper(curandStatePhilox4_32_10_t &state): state(state) {}
   __device__ float operator()() {
-    return curand_uniform(&state);
+    auto val = curand_uniform(&state);
+    // val is from 0.0 to 1.0, where 1.0 is included and 0.0 is excluded
+    // We want 1.0 excluded and 0.0 included
+    if (val == 1.0F) {
+        val = 0.0F;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Distributions.cu:76]: now you are oversampling 0s, severely, due to some curand quirks (curand divides uint by 2^32, so what would happen here is that 0 will be generated with the probability of 2^-24, because that's how often 1 is generated, but then 2^-32 will be generated with the probability 2^-32. The way to make it consistent with the rest of pytorch is to use transformation similar to the one in TransformationHelper that I linked (which generates 0, 2^-24, 2^-23 .... 1-2^24  with the same probability).

File: aten/src/ATen/native/Distributions.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/Distributions.h]: There are no guarantees that underlying generation techniques would not change, e.g. when the whole algorithm for generating Binomial changed it was ok.

File: aten/src/ATen/native/cuda/Distributions.cu (hunk 2 above)
Comment (mattip) [reply aten/src/ATen/native/cuda/Distributions.cu]: I would have thought that kind of mistake would have been caught by tests. Is this a known problem with `curand_uniform()`? In any case I adopted the code you linked to.

**Comment by mattip** [comment]
A few of the builds got this error, should I rebase off master again?
```
fatal: reference is not a tree: f015d698006c4a11be15b1ebb75b3b9bb317b914
Unable to checkout 'f015d698006c4a11be15b1ebb75b3b9bb317b914' in submodule \
    path 'third_party/tensorpipe'
```

File: aten/src/ATen/native/cuda/Distributions.cu [hunk 3]
```diff
@@ -69,7 +69,14 @@ struct curand_uniform_wrapper {
   curandStatePhilox4_32_10_t &state;
   __device__ curand_uniform_wrapper(curandStatePhilox4_32_10_t &state): state(state) {}
   __device__ float operator()() {
-    return curand_uniform(&state);
+    auto val = curand_uniform(&state);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Distributions.cu:72]: This won't work, because `val` is float, and you are static_casting and not reinterpret-casting it to uint64_t, which means that it'll either become 0 or 1. 
```suggestion
    auto val = curand_uniform(&state);
```
    uint val = curand(&state); //need just bits
    constexpr auto MASK = static_cast<uint>((static_cast<uint64_t>(1) << std::numeric_limits<float>::digits) - 1);//MASK should be uint, int64 arithmetic is much slower. In honesty, even static_cast<uint64_t>(1) is not needed because digist for float is 24, so 32 bits is enough, but it does not matter 
    constexpr auto DIVISOR = static_cast<float>(1) / (static_cast<uint>(1) << std::numeric_limits<float>::digits);//uint64_t is not needed here either
  return (val & MASK) * DIVISOR;
```

Comment (mattip) [reply aten/src/ATen/native/cuda/Distributions.cu:72]: thanks. I assumed tests would catch any gross errors :(

Comment (mattip) [reply aten/src/ATen/native/cuda/Distributions.cu]: adopted in 7257949. Also rebased off master to hopefully clear the CI errors

Comment (ngimel) [reply aten/src/ATen/native/cuda/Distributions.cu]: IT's actually concerning that the tests don't catch those errors.

Comment (ngimel) [reply aten/src/ATen/native/cuda/Distributions.cu]: It's uint32_t probably to fix windows builds, sorry.

**Comment by mattip** [comment]
The [`pr/pytorch-linux-xenial-rocm3.5.1-py3.6 ` build](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-rocm3.5.1-py3.6-test2/3840/console) is failing `test_sum_fp16 (__main__.TestCuda)` and `test_trilu_indices (__main__.TestCuda)`. I can't see a connection to my changes:
- this is cuda code, my changes are CPU only
- there should be no connection to binomial sampling in those tests

**Comment by ngimel** [comment]
You changes are cuda :-) but it does not look like they are related. ROCm has been very flaky lately. Let's try to merge.

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
Oh, btw, sorry to ask, but in the test you are doing, can you also roughly check the number of 0's and 1's, so that at least glaring errors in the distribution are caught?

**Comment by mattip** [comment]
> in the test you are doing, can you also roughly check the number of 0's and 1's

Yup.

**Comment by mattip** [comment]
I added a smoke test, but now there are new CI failures.

**Comment by ngimel** [comment]
Test failures are unrelated.
